{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.output {flex-direction: row;}</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 148,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import itertools\n",
    "from collections import Counter\n",
    "from IPython.display import display, HTML\n",
    "\n",
    "HTML('<style>.output {flex-direction: row;}</style>')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### toy example: count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 342,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# intersection of all expression vertices between cohort-level CLOvE pairs and the pairs in a sample (with that specific context)\n",
    "cohort = pd.DataFrame(np.random.randint(0,10,size=(16, 1)), columns=['score'])\n",
    "cohort['exp'] = ['a','b','c','d','a','b','c','d','a','b','c','d','a','b','c','d']\n",
    "cohort['cnv'] = ['A','A','A','A','B','B','B','B','C','C','C','C','D','D','D','D']\n",
    "\n",
    "cnv_s = pd.DataFrame(np.random.randint(0,2,size=(4, 4)), columns=['p1','p2','p3','p4'])\n",
    "exp_s = pd.DataFrame(np.random.rand(4, 4), columns=['p1','p2','p3','p4'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 343,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>p1</th>\n",
       "      <th>p2</th>\n",
       "      <th>p3</th>\n",
       "      <th>p4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.903135</td>\n",
       "      <td>0.221591</td>\n",
       "      <td>0.727339</td>\n",
       "      <td>0.617464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.207924</td>\n",
       "      <td>0.028553</td>\n",
       "      <td>0.991886</td>\n",
       "      <td>0.289566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.874191</td>\n",
       "      <td>0.064179</td>\n",
       "      <td>0.047417</td>\n",
       "      <td>0.648652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.336999</td>\n",
       "      <td>0.585830</td>\n",
       "      <td>0.146329</td>\n",
       "      <td>0.127086</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         p1        p2        p3        p4\n",
       "0  0.903135  0.221591  0.727339  0.617464\n",
       "1  0.207924  0.028553  0.991886  0.289566\n",
       "2  0.874191  0.064179  0.047417  0.648652\n",
       "3  0.336999  0.585830  0.146329  0.127086"
      ]
     },
     "execution_count": 343,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exp_s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 352,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "exp_s = exp_s[exp_s > 0.4]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 353,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\test\\Anaconda3\\lib\\site-packages\\ipykernel\\__main__.py:8: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    }
   ],
   "source": [
    "cnv_s['genes'] = ['A','B','C','D']\n",
    "cnv_s.set_index(keys='genes', drop=True,  inplace=True)\n",
    "\n",
    "exp_s['genes'] = ['a','b','c','d']\n",
    "exp_s.set_index(keys='genes', drop=True,  inplace=True)\n",
    "\n",
    "hits_s = cohort[cohort['score'] > 4]  # models significance cutoff\n",
    "hits_s['pair'] = hits_s['exp'] + hits_s['cnv']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 354,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>exp</th>\n",
       "      <th>cnv</th>\n",
       "      <th>pair</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9</td>\n",
       "      <td>a</td>\n",
       "      <td>A</td>\n",
       "      <td>aA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>8</td>\n",
       "      <td>b</td>\n",
       "      <td>A</td>\n",
       "      <td>bA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>5</td>\n",
       "      <td>d</td>\n",
       "      <td>A</td>\n",
       "      <td>dA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9</td>\n",
       "      <td>a</td>\n",
       "      <td>B</td>\n",
       "      <td>aB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>5</td>\n",
       "      <td>b</td>\n",
       "      <td>B</td>\n",
       "      <td>bB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>7</td>\n",
       "      <td>c</td>\n",
       "      <td>B</td>\n",
       "      <td>cB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>6</td>\n",
       "      <td>b</td>\n",
       "      <td>C</td>\n",
       "      <td>bC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>6</td>\n",
       "      <td>c</td>\n",
       "      <td>C</td>\n",
       "      <td>cC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>8</td>\n",
       "      <td>d</td>\n",
       "      <td>C</td>\n",
       "      <td>dC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>5</td>\n",
       "      <td>b</td>\n",
       "      <td>D</td>\n",
       "      <td>bD</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>6</td>\n",
       "      <td>d</td>\n",
       "      <td>D</td>\n",
       "      <td>dD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    score exp cnv pair\n",
       "0       9   a   A   aA\n",
       "1       8   b   A   bA\n",
       "3       5   d   A   dA\n",
       "4       9   a   B   aB\n",
       "5       5   b   B   bB\n",
       "6       7   c   B   cB\n",
       "9       6   b   C   bC\n",
       "10      6   c   C   cC\n",
       "11      8   d   C   dC\n",
       "13      5   b   D   bD\n",
       "15      6   d   D   dD"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>p1</th>\n",
       "      <th>p2</th>\n",
       "      <th>p3</th>\n",
       "      <th>p4</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genes</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>B</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>D</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       p1  p2  p3  p4\n",
       "genes                \n",
       "A       1   0   0   1\n",
       "B       1   1   0   1\n",
       "C       0   1   0   0\n",
       "D       0   0   1   0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>p1</th>\n",
       "      <th>p2</th>\n",
       "      <th>p3</th>\n",
       "      <th>p4</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genes</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>a</th>\n",
       "      <td>0.903135</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.727339</td>\n",
       "      <td>0.617464</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.991886</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c</th>\n",
       "      <td>0.874191</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.648652</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.58583</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             p1       p2        p3        p4\n",
       "genes                                       \n",
       "a      0.903135      NaN  0.727339  0.617464\n",
       "b           NaN      NaN  0.991886       NaN\n",
       "c      0.874191      NaN       NaN  0.648652\n",
       "d           NaN  0.58583       NaN       NaN"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "display(hits_s)\n",
    "display(cnv_s)\n",
    "display(exp_s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 350,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "p1\n",
      "['aA', 'aB', 'cA', 'cB', 'dA', 'dB']\n",
      "['d', 'a', 'a', 'c'] \n",
      "\n",
      "p2\n",
      "['dB', 'dC']\n",
      "['d'] \n",
      "\n",
      "p3\n",
      "['aD', 'bD']\n",
      "['b'] \n",
      "\n",
      "p4\n",
      "['aA', 'aB', 'cA', 'cB']\n",
      "['a', 'a', 'c'] \n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>p1</th>\n",
       "      <th>p2</th>\n",
       "      <th>p3</th>\n",
       "      <th>p4</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genes</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>a</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       p1  p2  p3  p4\n",
       "genes                \n",
       "a       2   0   0   2\n",
       "b       0   0   1   0\n",
       "c       1   0   0   1\n",
       "d       1   1   0   0"
      ]
     },
     "execution_count": 350,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vulnerability_vector_count(exp_s, cnv_s, hits_s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 360,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>p1</th>\n",
       "      <th>p2</th>\n",
       "      <th>p3</th>\n",
       "      <th>p4</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genes</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>a</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>c</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>d</th>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       p1  p2  p3  p4\n",
       "genes                \n",
       "a       2   0   0   2\n",
       "b       0   0   1   0\n",
       "c       1   0   0   1\n",
       "d       0   1   0   0"
      ]
     },
     "execution_count": 360,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vulnerability_vector_count(exp_s, cnv_s, hits_s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 363,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def vulnerability_vector_count(exp, cnv, hits):\n",
    "    \"\"\"Counts meaningful clDEGs (copy-loss differential expression genes) from loss contexts\n",
    "    \n",
    "    :param exp: df, gene x sample microarray or RNAseq, nan where no expression counts exist\n",
    "    :param cnv: df, gene x sample calls from GISTIC, 0=WT 1=loss\n",
    "    :param hits: df, pair x calculations from clove output, cutoff to sig, with 'pair' col\n",
    "    \n",
    "    returns gene x sample df of counts of significant expression linkages\"\"\"\n",
    "    \n",
    "    pat_essential = pd.DataFrame(0, index=exp.index, columns=exp.columns)\n",
    "    \n",
    "    for col in exp.columns:\n",
    "        \n",
    "        g_exp = exp[col].dropna().index.tolist()  # non-NaN genes from exp matrix\n",
    "        g_cnv = cnv[cnv[col] == 1][col].index.tolist()  # genes with value 1 from cnv matrix\n",
    "        \n",
    "        df = hits.loc[(hits['exp'].isin(g_exp)) & (hits['cnv'].isin(g_cnv))]\n",
    "        df = pd.DataFrame(df['exp'].value_counts())\n",
    "        \n",
    "        if df.shape[0] != 0:\n",
    "            df.columns=[col]\n",
    "            pat_essential[col] = df\n",
    "            \n",
    "    pat_essential.replace(np.nan, 0, inplace=True)\n",
    "    return pat_essential.astype(np.int32)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 317,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def vulnerability_vector_count_himem(exp, cnv, hits):\n",
    "    \"\"\"Counts meaningful gene expression linkages from loss contexts\n",
    "    \n",
    "    :param exp: df, gene x sample microarray or RNAseq, nan where no expression counts exist\n",
    "    :param cnv: df, gene x sample calls from GISTIC, 0=WT 1=loss\n",
    "    :param hits: df, pair x calculations from clove output, cutoff to sig, with 'pair' col\n",
    "    \n",
    "    returns gene x sample df of counts of significant expression linkages\"\"\"\n",
    "    \n",
    "    pat_essential = pd.DataFrame(0, index=exp.index, columns=exp.columns)\n",
    "    for col in exp.columns:\n",
    "        g_exp = exp[col].dropna().index.tolist()  # non-NaN genes from exp matrix\n",
    "        g_cnv = cnv[cnv[col] == 1][col].index.tolist()  # genes with value 1 from cnv matrix\n",
    "        \n",
    "        # building all combinations of pairs from exp and cnv is not needed (Mem Err)\n",
    "        pairs_list = [''.join(pair) for pair in itertools.product(g_exp, g_cnv)]  # form column as joined pair\n",
    "        essential = [g[0] for g in set(hits['pair']).intersection(pairs_list)]\n",
    "        print(col)\n",
    "        print(pairs_list)\n",
    "        print(essential,'\\n')\n",
    "        \n",
    "        # try instead to pull from hits['pair']\n",
    "        df = pd.DataFrame.from_dict(dict(Counter(essential)), orient='index')\n",
    "        if df.shape[0] != 0:\n",
    "            df.columns=[col]\n",
    "            pat_essential[col] = df\n",
    "    pat_essential.replace(np.nan, 0, inplace=True)\n",
    "    return pat_essential.astype(np.int32)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 281,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def prepare_vv(exp, cnv, cloves, sig=0.01):\n",
    "    \"\"\"Prepare CLoVE computations df for vulnerability vector\n",
    "    \n",
    "    :param cloves: df, pair x calculations from clove output\n",
    "    :param exp: df, gene x sample microarray or RNAseq, nan where no expression counts exist\n",
    "    :param cnv: df, gene x sample calls from GISTIC, 0=WT 1=loss\n",
    "    :param sig: float, p-value threshold for signifiance, default=0.01\n",
    "    \n",
    "    returns cutoff'd pair x calculations df from clove output with 'pair' col\n",
    "    \"\"\"\n",
    "    \n",
    "    if 'pair' not in cloves.columns:\n",
    "        cloves['pair'] = cloves['exp'] + cloves['cnv']\n",
    "#         for df in [exp, cnv]:\n",
    "#             if df.index.dtype != 'object':\n",
    "#                 df.set_index(keys='gene_id', drop=True, inplace=True)\n",
    "    if exp.index.dtype != 'object':\n",
    "        exp.set_index(keys='gene_id', drop=True, inplace=True)\n",
    "    if cnv.index.dtype != 'object':\n",
    "        cnv.set_index(keys='gene_id', drop=True, inplace=True)\n",
    "    \n",
    "    return exp, cnv, cloves[cloves['np_p_w'] <= sig]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 361,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "exp = pd.read_csv('brca_exp.tab.gz', sep='\\t', compression='gzip')\n",
    "cnv = pd.read_csv('brca_cnv_het.tab.gz', sep='\\t', compression='gzip')\n",
    "scores = pd.read_csv('brca_1M_clove.tab.gz', sep='\\t', compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 364,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "exp, cnv, scores = prepare_vv(exp, cnv, scores, sig=0.001)\n",
    "counts = vulnerability_vector_count(exp, cnv, scores)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 375,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "counts.to_csv('brca_het_clDEG_count.tab.gz', sep='\\t', compression='gzip')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 367,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-AR-A24V-01A</th>\n",
       "      <th>TCGA-A7-A26H-01A</th>\n",
       "      <th>TCGA-EW-A6SB-01A</th>\n",
       "      <th>TCGA-C8-A137-01A</th>\n",
       "      <th>TCGA-S3-A6ZG-01A</th>\n",
       "      <th>TCGA-BH-A42T-01A</th>\n",
       "      <th>TCGA-OL-A5RU-01A</th>\n",
       "      <th>TCGA-AQ-A54N-01A</th>\n",
       "      <th>TCGA-AO-A0J3-01A</th>\n",
       "      <th>TCGA-A7-A5ZV-01A</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-BH-A18L-01A</th>\n",
       "      <th>TCGA-AO-A1KT-01A</th>\n",
       "      <th>TCGA-AR-A24S-01A</th>\n",
       "      <th>TCGA-AO-A1KS-01A</th>\n",
       "      <th>TCGA-GM-A3XN-01A</th>\n",
       "      <th>TCGA-A2-A25E-01A</th>\n",
       "      <th>TCGA-E2-A15O-01A</th>\n",
       "      <th>TCGA-AR-A5QP-01A</th>\n",
       "      <th>TCGA-E9-A24A-01A</th>\n",
       "      <th>TCGA-AN-A0FJ-01A</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>A1BG</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A1CF</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2BP1</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2LD1</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>A2M</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1066 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         TCGA-AR-A24V-01A  TCGA-A7-A26H-01A  TCGA-EW-A6SB-01A  \\\n",
       "gene_id                                                         \n",
       "A1BG                    0                 0                 0   \n",
       "A1CF                    0                 0                 1   \n",
       "A2BP1                   0                 0                 0   \n",
       "A2LD1                   2                 2                 4   \n",
       "A2M                     0                 0                 2   \n",
       "\n",
       "         TCGA-C8-A137-01A  TCGA-S3-A6ZG-01A  TCGA-BH-A42T-01A  \\\n",
       "gene_id                                                         \n",
       "A1BG                    0                 0                 0   \n",
       "A1CF                    1                 0                 1   \n",
       "A2BP1                   0                 0                 0   \n",
       "A2LD1                   3                 0                 1   \n",
       "A2M                     0                 1                 1   \n",
       "\n",
       "         TCGA-OL-A5RU-01A  TCGA-AQ-A54N-01A  TCGA-AO-A0J3-01A  \\\n",
       "gene_id                                                         \n",
       "A1BG                    0                 0                 0   \n",
       "A1CF                    0                 2                 2   \n",
       "A2BP1                   0                 0                 0   \n",
       "A2LD1                   0                 4                 3   \n",
       "A2M                     0                 2                 2   \n",
       "\n",
       "         TCGA-A7-A5ZV-01A        ...         TCGA-BH-A18L-01A  \\\n",
       "gene_id                          ...                            \n",
       "A1BG                    0        ...                        0   \n",
       "A1CF                    0        ...                        0   \n",
       "A2BP1                   0        ...                        0   \n",
       "A2LD1                   0        ...                        0   \n",
       "A2M                     0        ...                        1   \n",
       "\n",
       "         TCGA-AO-A1KT-01A  TCGA-AR-A24S-01A  TCGA-AO-A1KS-01A  \\\n",
       "gene_id                                                         \n",
       "A1BG                    0                 0                 0   \n",
       "A1CF                    0                 2                 0   \n",
       "A2BP1                   0                 0                 0   \n",
       "A2LD1                   1                 3                 1   \n",
       "A2M                     0                 2                 1   \n",
       "\n",
       "         TCGA-GM-A3XN-01A  TCGA-A2-A25E-01A  TCGA-E2-A15O-01A  \\\n",
       "gene_id                                                         \n",
       "A1BG                    0                 0                 0   \n",
       "A1CF                    0                 1                 0   \n",
       "A2BP1                   0                 0                 0   \n",
       "A2LD1                   0                 1                 0   \n",
       "A2M                     1                 0                 1   \n",
       "\n",
       "         TCGA-AR-A5QP-01A  TCGA-E9-A24A-01A  TCGA-AN-A0FJ-01A  \n",
       "gene_id                                                        \n",
       "A1BG                    0                 0                 0  \n",
       "A1CF                    0                 0                 2  \n",
       "A2BP1                   0                 0                 0  \n",
       "A2LD1                   0                 0                 3  \n",
       "A2M                     1                 1                 2  \n",
       "\n",
       "[5 rows x 1066 columns]"
      ]
     },
     "execution_count": 367,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 371,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene_id</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CDK1</th>\n",
       "      <td>14148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TROAP</th>\n",
       "      <td>13866</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>STIP1</th>\n",
       "      <td>13756</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>H2AFZ</th>\n",
       "      <td>13664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GINS1</th>\n",
       "      <td>13319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCNB1</th>\n",
       "      <td>13044</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CDC45</th>\n",
       "      <td>13016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CENPI</th>\n",
       "      <td>12877</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KIAA0247</th>\n",
       "      <td>12772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>JMJD6</th>\n",
       "      <td>12748</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>UBXN10</th>\n",
       "      <td>12570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PLK1</th>\n",
       "      <td>12403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CENPE</th>\n",
       "      <td>12108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FOXM1</th>\n",
       "      <td>12106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RRM2</th>\n",
       "      <td>12071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CIRBP</th>\n",
       "      <td>12007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TUBA1C</th>\n",
       "      <td>11869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>C16orf57</th>\n",
       "      <td>11839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ALG3</th>\n",
       "      <td>11751</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HNRNPAB</th>\n",
       "      <td>11694</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              0\n",
       "gene_id        \n",
       "CDK1      14148\n",
       "TROAP     13866\n",
       "STIP1     13756\n",
       "H2AFZ     13664\n",
       "GINS1     13319\n",
       "CCNB1     13044\n",
       "CDC45     13016\n",
       "CENPI     12877\n",
       "KIAA0247  12772\n",
       "JMJD6     12748\n",
       "UBXN10    12570\n",
       "PLK1      12403\n",
       "CENPE     12108\n",
       "FOXM1     12106\n",
       "RRM2      12071\n",
       "CIRBP     12007\n",
       "TUBA1C    11869\n",
       "C16orf57  11839\n",
       "ALG3      11751\n",
       "HNRNPAB   11694"
      ]
     },
     "execution_count": 371,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.DataFrame(counts.sum(axis=1).sort_values(ascending=False)).head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 373,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(20502, 1066)"
      ]
     },
     "execution_count": 373,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts.shape\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 374,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(160654, 16)"
      ]
     },
     "execution_count": 374,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scores.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
